### ISSUE TITLE
Recommended approach to adding assembly search directories when using AssemblyLoadContext.LoadFromAssemblyPath?

### ISSUE THREAD

**Issue by nathan-alden-hp** [dotnet/core#3090]
I'm running .NET Core 3.0 preview 7. Referencing this issue for somewhat related context: https://github.com/dotnet/core/issues/290.

I am writing a plugin architecture using a .NET Core 3.0 WPF application and .NET Standard 2.0 class libraries for the plugins. In my architecture, the main application loads plugins from a user-configured directory. The code uses `AssemblyLoadContext.LoadFromAssemblyPath` to accomplish this. It works perfectly as long as the plugin assembly has no dependencies on .NET assemblies in the same directory.

In my contrived example, my plugin references the `System.Management` NuGet package version `4.6.0-preview7.19362.9`. In the plugin project file, I included `<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>` to get the compiler to copy `System.Management.dll` to the output directory. My WPF application then attempts to load my plugin assembly from this output directory. The plugin assembly loads, but as soon as I invoke code that tries to use `System.Management`, it throws an exception stating it can't find `System.Management`.

I then went looking for a way to instruct the assembly resolver to include a specific directory in its search, but unfortunately this does not seem to be possible. The recommendation appears to be to use the `AssemblyLoadContext.Resolving` event, which is being called for assembly resolutions that fail. Unfortunately, the `AssemblyName` parameter in the delegate contains the qualified assembly name and not the DLL filename. I am not sure how to (or if I should) translate this qualified name into a filename, check that the version number and culture match, etc.

![image](https://user-images.githubusercontent.com/40613229/61841941-6ff80a80-ae64-11e9-93c7-2d8ee0c7b8fe.png)

An additional potential issue is that many other assemblies--that may or may not exist (e.g., `MyPlugin.resources.dll`)--are also triggering `AssemblyLoadContext.Resolving` events and I can't find a way to detect which are "real" references and which aren't. I define "real" to mean "assemblies referenced with `.assembly extern` in the IL of the plugin assembly":

```
.assembly extern System.Management
{
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )                         // .?_....:
  .ver 4:0:0:0
}
```

What is the appropriate way to solve this issue? FWIW, it would really be helpful to have a way to add to the list of directories being searched during assembly resolution, as performing qualified name-to-filename mappings, version checking, etc. seems fraught with potential issues.

**Comment by nathan-alden-hp** [comment]
Possible answer to my question (although it really shouldn't require this much code): https://samcragg.wordpress.com/2017/06/30/resolving-assemblies-in-net-core/

**Comment by nathan-alden-hp** [comment]
I have yet to find success with the blog article I posted above. Unfortunately, I get to this line:
```
RuntimeLibrary runtimeLibrary = _dependencyContext.RuntimeLibraries.FirstOrDefault(a => string.Equals(a.Name, name.Name, StringComparison.OrdinalIgnoreCase));
```
and it doesn't find a matching runtime library. I compared the contents of the `"libraries"` property in my plugin assembly's `.deps.json` file and `RuntimeLibraries` is missing `System.Management` while `MyProject.Plugin.deps.json` has it.

`RuntimeLibraries`:
```
Microsoft.NETCore.App.Internal
Microsoft.NETCore.DotNetHostPolicy
Microsoft.NETCore.Platforms
Microsoft.NETCore.Targets
Microsoft.WindowsDesktop.App
MyProject.Plugin
MyProject.Sdk
NETStandard.Library
runtime.win-x64.Microsoft.NETCore.App
runtime.win-x64.Microsoft.NETCore.DotNetHostPolicy
runtime.win-x64.Microsoft.WindowsDesktop.App
Serilog
System.Collections.NonGeneric
System.Diagnostics.Debug
System.Globalization
System.IO
System.Reflection
System.Reflection.Primitives
System.Resources.ResourceManager
System.Runtime
System.Runtime.Extensions
System.Text.Encoding
System.Threading
System.Threading.Tasks
```
`MyProject.Plugin.deps.json`:
```
"libraries": {
  "MyProject.Plugin/1.0.0": {...},
  "Microsoft.NETCore.App/3.0.0-preview7-27912-14": {...},
  "Microsoft.NETCore.Platforms/3.0.0-preview7.19362.9": {...},
  "Microsoft.NETCore.Targets/1.1.0": {...},
  "Microsoft.Win32.Registry/4.6.0-preview7.19362.9": {...},
  "Microsoft.WindowsDesktop.App/3.0.0-preview7-27912-14": {...},
  "Serilog/2.8.0": {...},
  "System.CodeDom/4.6.0-preview7.19362.9": {...},
  "System.Collections.NonGeneric/4.3.0": {...},
  "System.Diagnostics.Debug/4.3.0": {...},
  "System.Globalization/4.3.0": {...},
  "System.IO/4.3.0": {...},
  "System.Management/4.6.0-preview7.19362.9": {...},
  "System.Reflection/4.3.0": {...},
  "System.Reflection.Primitives/4.3.0": {...},
  "System.Resources.ResourceManager/4.3.0": ...},
  "System.Runtime/4.3.0": {...},
  "System.Runtime.Extensions/4.3.0": {...},
  "System.Runtime.Loader/4.3.0": {...},
  "System.Security.AccessControl/4.6.0-preview7.19362.9": {...},
  "System.Security.Principal.Windows/4.6.0-preview7.19362.9": {...},
  "System.Text.Encoding/4.3.0": {...},
  "System.Threading/4.3.0": {...},
  "System.Threading.Tasks/4.3.0": {...},
  "MyProject.Sdk/1.0.0": {...}
}
```
Is this because `System.Management` hasn't been loaded yet? Why won't `DependencyContext` pick up the `System.Management` dependency?

**Comment by nathan-alden-hp** [comment]
I identified a race condition in my code that was causing this issue caused by attempting to load several plugins. In the above blog's code, the author attaches an event handler to `AssemblyLoadContext.Resolving`, but his code assumes only one assembly will be loaded. Calling the author's code several times results in multiple event handlers being attached to `AssemblyLoadContext.Resolving`, thus messing up the underlying resolution algorithm.

When I changed my code to only load one plugin, the resolver finds `System.Management`. Unfortunately, there is now a separate issue:
>An exception occurred (System.PlatformNotSupportedException) retrieving system information: System.Management currently is only supported for Windows desktop applications.

I will close this issue as this `PlatformNotSupportedException` is a separate issue.
